### PR TITLE
Merge `dev` into `feat-fast-sync`.

### PR DESCRIPTION
**Pull request by afck**
Closes https://github.com/casper-network/casper-node/issues/2042.

Take a look at https://github.com/casper-network/casper-node/pull/2266 to see the changes compared to `dev` instead of `feat-fast-sync`.

### CODE REVIEW

File: .drone.yml [hunk 1]
```diff
@@ -9,6 +9,13 @@ name: pre-checks
 
 # Steps perform as fast serially, due to file thrashing.
 steps:
+
+    # TODO: re-enable once we figure out publishing
+    # - name: markdown-link-check
+    #   image: casperlabs/node-build-u1804
+    #   commands:
+    #     - "./ci/markdown_link_check.sh"
```
Comment (xcthulhu) [inline .drone.yml:17]: What happened here?

File: client/lib/ffi.rs [hunk 2]
```diff
@@ -25,7 +25,7 @@ fn set_last_error(error: Error) {
 ///
 /// The full error can be extracted with get_last_error. See [Error](super::Error) for more details
 /// on what these mean.
-#[allow(non_snake_case, non_camel_case_types, missing_docs)]
```
Comment (xcthulhu) [inline client/lib/ffi.rs:28]: I'm curious why we are changing this here in the ffi...?

File: execution_engine/CHANGELOG.md [hunk 3]
```diff
@@ -15,18 +15,26 @@ All notable changes to this project will be documented in this file.  The format
 
 ### Added
 * Added genesis validation step to ensure there are more genesis validators than validator slots.
+* Added a support for passing a public key as a `target` argument in native transfers.
+* Added a `max_associated_keys` configuration option for a hard limit of associated keys under accounts.
+
+
```
Comment (xcthulhu) [inline execution_engine/CHANGELOG.md:21]: Few too many whitespaces here

File: execution_engine/src/core/engine_state/executable_deploy_item.rs [hunk 4]
```diff
@@ -1,3 +1,4 @@
+//! Units of execution.
 // TODO - remove once schemars stops causing warning.
 #![allow(clippy::field_reassign_with_default)]
```
Comment (xcthulhu) [inline execution_engine/src/core/engine_state/executable_deploy_item.rs:3]: Can this be removed?

File: .drone.yml (hunk 1 above)
Comment (afck) [reply .drone.yml:17]: I don't know: https://github.com/casper-network/casper-node/commit/4e31be1c587220d6ebfffe04f16c22332b091070

File: execution_engine/CHANGELOG.md (hunk 3 above)
Comment (afck) [reply execution_engine/CHANGELOG.md:21]: Would be better to fix this on `dev`, not `feat-fast-sync`. We didn't change this file.

File: execution_engine/src/core/engine_state/executable_deploy_item.rs (hunk 4 above)
Comment (afck) [reply execution_engine/src/core/engine_state/executable_deploy_item.rs:3]: Also better on `dev` maybe.

File: client/lib/validation.rs [hunk 5]
```diff
@@ -4,11 +4,10 @@ use jsonrpc_lite::JsonRpc;
 use thiserror::Error;
```
Comment (xcthulhu) [inline client/lib/validation.rs:4]: Might as well merge this whole file back into `dev`

File: node/src/components/block_validator.rs [hunk 6]
```diff
@@ -203,7 +203,7 @@ where
             })
             .join(", ");
         info!(
-            ?sender, %duplicates,
+            peer_id=?sender, %duplicates,
             "received invalid block containing duplicated deploys"
         );
     }
```
Comment (xcthulhu) [inline node/src/components/block_validator.rs:209]: Below in this file we should remove:

```rust
        if deploy.deploy_or_transfer_hash() != dt_hash {
            warn!(
                deploy = ?deploy,
                expected_deploy_or_transfer_hash = ?dt_hash,
                actual_deploy_or_transfer_hash = ?deploy.deploy_or_transfer_hash(),
                "Deploy has incorrect transfer hash"
            );
            return Event::CannotConvertDeploy(dt_hash);
        }
```

**Comment by afck** [comment]
I created a draft PR https://github.com/casper-network/casper-node/pull/2266 (that we won't merge!) which shows the diff with `dev` instead of `feat-fast-sync`.

**Review by xcthulhu — APPROVED** [review]

**Review by fizyk20 — APPROVED** [review]

**Comment by afck** [comment]
bors merge

**Comment by xcthulhu** [comment]
@afck I don't think we can do bors-merge because the upgrade tests are failing.

**Comment by afck** [comment]
Right, that's https://github.com/casper-network/casper-node/issues/1665 and not due to this merge. :+1:

**Pull request merged**